l1-contracts/src/core/libraries/decoders/TxsDecoder.sol

```diff
@@ -148,9 +154,38 @@ library TxsDecoder {
          * Compute encrypted and unencrypted logs hashes corresponding to the current leaf.
          * Note: will advance offsets by the number of bytes processed.
          */
-        (vars.noteEncryptedLogsHash, offset) = computeKernelLogsHash(offset, _body, true);
-        (vars.encryptedLogsHash, offset) = computeKernelLogsHash(offset, _body, false);
-        (vars.unencryptedLogsHash, offset) = computeKernelLogsHash(offset, _body, false);
+        offsets.encryptedLogsLength = offset;
+        offset += 0x20;
+        offsets.unencryptedLogsLength = offset;
+        offset += 0x20;
+
+        (vars.noteEncryptedLogsHash, offset, vars.kernelNoteEncryptedLogsLength) =
+          computeKernelLogsHash(offset, _body, true);
+        (vars.encryptedLogsHash, offset, vars.kernelEncryptedLogsLength) =
+          computeKernelLogsHash(offset, _body, false);
+        (vars.unencryptedLogsHash, offset, vars.kernelUnencryptedLogsLength) =
+          computeKernelLogsHash(offset, _body, false);
+
+        // We throw to ensure that the byte len we charge for DA gas in the kernels matches the actual chargable log byte len
+        // Without this check, the user may provide the kernels with a lower log len than reality
+        if (
+          uint256(bytes32(slice(_body, offsets.encryptedLogsLength, 0x20)))
```

Comment:
Why not having a separate length for note logs or combine all 3 together?

Reply:
I can't think of any reason that will require 3 lengths. But let's have a length for each type of log to begin with. It's easier to read. And we can probably refactor some repeated code and treat them as the same type { hash, length }.
Don't have to change it now if you want to merge this one today :)